{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d9d78e4b-19aa-41ef-95db-bbe9551e3d85",
   "metadata": {},
   "source": [
    "# Data Types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f4e88df-bc76-4ac7-9ce1-d65c46d412b1",
   "metadata": {},
   "source": [
    "## 1. Numeric"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e268770-5297-4a94-8646-a2748da49c3b",
   "metadata": {},
   "source": [
    "There are three numeric types in Python:\n",
    "\n",
    "- Int: Int, or integer, is a whole number, positive or negative, without decimals, of unlimited length.\n",
    "\n",
    "- Float: Float, or \"floating point number\" is a number, positive or negative, containing one or more decimals. Float can also be scientific numbers with an \"e\" to indicate the power of 10.\n",
    "\n",
    "- Complex: Complex numbers are written with a \"j\" as the imaginary part:\n",
    "\n",
    "Variables of numeric types are created when you assign a value to them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ac45ef2-e52c-414a-b4df-11a3dc93d06d",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = 1    # int\n",
    "y = 2.8  # float\n",
    "z = 1j   # complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2ac19f0a-d847-4061-91c7-22dd27448dca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of human genes: 21306\n",
      "Number of human genes in US: 7007934855138\n"
     ]
    }
   ],
   "source": [
    "human_genes = 21306         # int type. Not 21,306 or 21 306\n",
    "US_population = 328918373\n",
    "print('Number of human genes:', human_genes)\n",
    "print('Number of human genes in US:', human_genes*US_population)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aa14cdde-6caf-4de7-b500-9cfa54bb49b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "human exons per gene: 8.9\n",
      "human_exons = 189623.4\n",
      "Number of human exons: 189623.4\n"
     ]
    }
   ],
   "source": [
    "# Floats are implemented using double in C\n",
    "exons_per_gene = 8.9        # float type\n",
    "print('human exons per gene:', exons_per_gene)\n",
    "human_exons = exons_per_gene*human_genes\n",
    "print(\"human_exons =\", human_exons)\n",
    "print('Number of human exons:', human_exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "64682e24-7b33-45f9-a567-e4f8ca551f40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Approximate number of human exons: 189623\n"
     ]
    }
   ],
   "source": [
    "# To convert float to integer\n",
    "human_exons = int(human_exons)\n",
    "print('Approximate number of human exons:', human_exons)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a96e66b-fb79-42c0-a584-f5c4ded463ba",
   "metadata": {},
   "source": [
    "**convert from int to float:**\n",
    "a = float(x)\n",
    "\n",
    "**convert from float to int:**\n",
    "b = int(y)\n",
    "\n",
    "**convert from int to complex:**\n",
    "c = complex(x)\n",
    "\n",
    "print(a)\n",
    "print(b)\n",
    "print(c)\n",
    "\n",
    "print(type(a))\n",
    "print(type(b))\n",
    "print(type(c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d125c1b2-0cae-4843-89f9-7792e47e00e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9.4*0.2321)*5.6 - 9.4*(0.2321*5.6) = -1.7763568394002505e-15\n"
     ]
    }
   ],
   "source": [
    "# Can Python do arithmetic?\n",
    "import math\n",
    "firstProduct = (9.4*0.2321)*5.6\n",
    "secondProduct = 9.4*(0.2321*5.6)\n",
    "print('(9.4*0.2321)*5.6 - 9.4*(0.2321*5.6) =',\n",
    "      (firstProduct - secondProduct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3b93692d-b22a-4d89-a0f0-e84719bf0a47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "two_pi = 6.283185307179586\n",
      "sin(two_pi) = -2.4492935982947064e-16\n",
      "Do you believe this result?\n"
     ]
    }
   ],
   "source": [
    "# To access mathematical functions from math module\n",
    "two_pi = 2.0*math.pi\n",
    "print('two_pi =', two_pi)\n",
    "print('sin(two_pi) =', math.sin(two_pi))\n",
    "print('Do you believe this result?')\n",
    "#Mathematically, sin(2ùúã)= 0 but due to floating-point precision limits in computers, Python returns a very small number close to zero"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abce4caf-038c-4957-8aff-1218bcee4584",
   "metadata": {},
   "source": [
    "---\n",
    "## 2. Strings\n",
    "Strings in python are surrounded by either single quotation marks, or double quotation marks.\n",
    "\n",
    "'hello' is the same as \"hello\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7561b47f-f94a-4ada-b551-29245135c713",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein sequence of GFP: MSKGEELFTG...HGMDELYK\n"
     ]
    }
   ],
   "source": [
    "protein = \"GFP\"                    # Winner of 2008 Nobel in chemistry\n",
    "protein_seq_begin = 'MSKGEELFTG'\n",
    "protein_seq_end = 'HGMDELYK'\n",
    "\n",
    "# Concatenation of strings\n",
    "protein_seq = protein_seq_begin + '...' + protein_seq_end\n",
    "print('Protein sequence of GFP: ' + protein_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da4460a-84a2-4cdd-a248-8c390a47be9a",
   "metadata": {},
   "source": [
    "DNA_seq = DNA_seq.upper()\n",
    "\n",
    "- The **.upper() string method** converts all lowercase letters in the string to uppercase.\n",
    "\n",
    "Strings in Python are immutable (they can‚Äôt be changed in place), so .upper() returns a new string, that is reassigned back to DNA_seq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f6f69202-c362-416a-917b-118273879286",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA sequence: ATGAGTAAAG...ACTATACAAA\n"
     ]
    }
   ],
   "source": [
    "# String method str.upper()\n",
    "DNA_seq = 'atgagtaaag...actatacaaa'\n",
    "DNA_seq = DNA_seq.upper()\n",
    "print('DNA sequence: ' + DNA_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "df4d136b-7741-4f85-9197-4f8a62b6e791",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The second nucleotide: T\n",
      "The last nucleotide: A\n"
     ]
    }
   ],
   "source": [
    "# Forward index starts with 0 and increases\n",
    "# Backward index starts with -1 and decreases\n",
    "print('The second nucleotide:', DNA_seq[1])\n",
    "print('The last nucleotide:', DNA_seq[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0d7ebf17-cdd6-45db-a3b2-df997160a845",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "first codon: ATG\n",
      "Last codon: AAA\n"
     ]
    }
   ],
   "source": [
    "# Slicing a string\n",
    "first_codon = DNA_seq[0:3]         # index 3 excluded\n",
    "last_codon = DNA_seq[-3:]\n",
    "print(\"first codon:\", first_codon)\n",
    "print('Last codon:', last_codon)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdd847c-7f93-4ac6-ae58-b2d311ffff6f",
   "metadata": {},
   "source": [
    "---\n",
    "### 3. List\n",
    "Lists are used to store multiple items in a single variable.\n",
    "\n",
    "Lists are one of 4 built-in data types in Python used to store collections of data, the other 3 are Tuple, Set, and Dictionary, all with different qualities and usage.\n",
    "\n",
    "Lists are created using square brackets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8ecfd52e-3910-49da-832a-59487cd08cdc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TAA', 'tAG']\n"
     ]
    }
   ],
   "source": [
    "# Constructing a list\n",
    "stop_codons = ['TAA', 'tAG']\n",
    "print(stop_codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "40c7eb5d-fa1b-4d23-ac00-2ad3f432f923",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAA\n"
     ]
    }
   ],
   "source": [
    " # Accessing an item in a list\n",
    "first_stop_codon = stop_codons[0]\n",
    "print(first_stop_codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "51db7e7e-ef1e-4b4b-bb4a-eff54d0f1b57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TAA', 'TAG']\n"
     ]
    }
   ],
   "source": [
    "# Modifying an item in a list\n",
    "stop_codons[1] = 'TAG'\n",
    "print(stop_codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4c5d7ed4-3fb0-4d90-b7f0-d6a29657e01c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TAA', 'TAG', 'TGA']\n"
     ]
    }
   ],
   "source": [
    " # Appending an item to the end of a list\n",
    "stop_codons.append('TGA')\n",
    "print(stop_codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0089bd96-4e50-474a-b7a8-e4785c3c68c2",
   "metadata": {},
   "source": [
    "**stop_codons.append('TGA')**\n",
    "\n",
    "The .append() method adds a new element to the end of the list.  \n",
    "Note: .append() modifies the original list in place, it doesn‚Äôt create a new list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "98748e40-8e56-4c42-8f7d-d2ca9d2d6783",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3 stop codons\n"
     ]
    }
   ],
   "source": [
    "# Number of items in a list\n",
    "number_of_stop_codons = len(stop_codons)\n",
    "print('There are', number_of_stop_codons, 'stop codons')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbc348a8-14c4-467f-89c2-5aa595254a1d",
   "metadata": {},
   "source": [
    "**len** for length of a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "590137bd-aa28-4262-ab0d-0ae6b9ae8719",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3 stop codons\n"
     ]
    }
   ],
   "source": [
    "# Number of items in a list\n",
    "number_of_stop_codons = len(stop_codons)\n",
    "print('There are', number_of_stop_codons, 'stop codons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5060420f-9ed1-466c-9f39-afb6d35cd376",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAATAGTGA\n"
     ]
    }
   ],
   "source": [
    "# Convert list to a string\n",
    "DNA_seq = ''.join(stop_codons)\n",
    "print(DNA_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efbb5ab8-5977-4890-8428-8c8d389f0c39",
   "metadata": {},
   "source": [
    "The join() method combines all elements of a list into a single string.\n",
    "\n",
    "The string before .join() ( ' ' in this case) is the separator placed between each element.\n",
    "\n",
    "Since the separator is an empty string (''), all list elements are joined without spaces or characters between them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3f73f613-56e7-4835-9bc7-92cde3db9aed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['T', 'A', 'A', 'T', 'A', 'G', 'T', 'G', 'A']\n"
     ]
    }
   ],
   "source": [
    "# Convert string to a list\n",
    "DNA_list = list(DNA_seq)\n",
    "print(DNA_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "83d77407-a5ea-4189-b46f-b7c230690ab4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Second codon: ['T', 'A', 'G']\n"
     ]
    }
   ],
   "source": [
    "# Slicing a list\n",
    "second_codon = DNA_list[3:6]             # index 6 not included\n",
    "print('Second codon:', second_codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f183f598-a48e-461b-8281-31a6c003c54a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['T', 'A', 'A', 'T', 'A', 'G', 'T', 'G', 'A']\n"
     ]
    }
   ],
   "source": [
    "# Copying a list\n",
    "DNA_list_duplicate = DNA_list.copy()\n",
    "print(DNA_list_duplicate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "855da5f6-7f5e-44fd-9dcc-b3cb7d102141",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['T', 'A', 'A', 'T', 'A', '?', 'G', 'T', 'G', 'A']\n",
      "['T', 'A', 'A', 'T', 'A', 'G', 'T', 'G', 'A']\n"
     ]
    }
   ],
   "source": [
    "# Insert, delete element\n",
    "DNA_list_duplicate.insert(5, \"?\") #he \"?\" is inserted before the element that was originally at index 5.\n",
    "print(DNA_list_duplicate) \n",
    "\n",
    "DNA_list_duplicate.pop(5)        # Can also use: del DNA_list_duplicate[5]. The .pop here removes the element at index 5 (which is \"?\").\n",
    "print(DNA_list_duplicate)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95b364a7-a695-4e6b-abab-d7232d339961",
   "metadata": {},
   "source": [
    "**list.pop(index)** or **del** :If no index is given, it removes the last element by default."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c9716b0-84ef-4a01-b06c-358c53523a8b",
   "metadata": {},
   "source": [
    "---\n",
    "## Tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e3be6dc-9474-47f3-9a84-5622619c9d0e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
